verilog/analysis/checkers/disable_fork_rule.cc

```diff
@@ -0,0 +1,118 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/disable_fork_rule.h"
+
+#include <set>
+#include <string>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/analysis/matcher/core_matchers.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/CST/identifier.h"
+#include "verilog/CST/verilog_matchers.h"  // IWYU pragma: keep
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::SyntaxTreeContext;
+using verible::matcher::Matcher;
+
+// Register DisableForkNoLabelsRule
+VERILOG_REGISTER_LINT_RULE(DisableForkNoLabelsRule);
+
+absl::string_view DisableForkNoLabelsRule::Name() {
+  return "disable-statement";
+}
+const char DisableForkNoLabelsRule::kTopic[] = "disable-statement";
+const char DisableForkNoLabelsRule::kMessage[] =
+    "Invalid usage of disable statement. Allowed construction is: disable "
+    "fork;";
+
+std::string DisableForkNoLabelsRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat("Checks that there are no occurrences of ",
+                      Codify("disable some_label", description_type), ". The ",
+                      Codify("disable some_label", description_type),
+                      " can point to the labels created with begin keyword "
+                      "without initial/final/always specifier",
+                      ". Use ", Codify("disable fork", description_type),
+                      " instead. See ", GetStyleGuideCitation(kTopic), ".");
+}
+
+static const Matcher& DisableMatcher() {
+  static const Matcher matcher(
+      NodekDisableStatement(DisableStatementHasLabel()));
+  return matcher;
+}
+
+void DisableForkNoLabelsRule::HandleSymbol(const verible::Symbol& symbol,
+                                           const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (DisableMatcher().Matches(symbol, &manager)) {
+    // if no kDisable label, return, nothing to be checked
+    const auto disableLabels = FindAllSymbolIdentifierLeafs(symbol);
+    if (disableLabels.size() == 0) {
+      return;
+    }
+    // validate all kBegin parents
```

Comment:
In general, did you a what `context` has to offer ? There are all sorts of utility functions like `DirectParentsAre()` or `IsInside()` that can help formulate what you're looking for more specifically.

Reply:
Yes, I was there. Unfortunately, I couldn't use any of these functions, cause I need to retrieve the parent nodes to obtain the node's labels